### PR TITLE
Allow to connect without resetting.

### PR DESCRIPTION
**Pull request by baffo32**
This patch allows the user to connect and disconnect from the 32-bit board without resetting it or changing the streaming state.

This is useful to connect to the board without interrupting SD logging.  This would also be invaluable for a CLI version of the API, so tools could engage the board state without having to be responsible for it.

This obviously wouldn't work for the 8-bit board, which I do not have to test with.  Does this library support the 8-bit board?

### CODE REVIEW

**Comment by baffo32** [comment]
Whoops, failing test.  Guess this needs more work.  Withdrawn for now.

**Comment by aj-ptw** [comment]
Run tests with out a dongle plugged in with `npm test`

**Comment by aj-ptw** [comment]
Also I think we should make that arg optional and default to the current functionality to prevent a breaking change.

**Comment by baffo32** [comment]
Tests pass now.  The arg is indeed optional, and defaults to the current functionality.  When not included it will be `null` and hence evaluate to `false` in tests.

**Comment by aj-ptw** [comment]
Did you add coverage for these? If you need help writing the tests I will happily help!!

**Comment by baffo32** [comment]
Sorry, I have not added tests!

Here's my notes.  If you find this easy by all means jump on it.  Otherwise I'll learn the testing API eventually.

When a second parameter of `true` is passed to connect(), no commands shall be written to the board, the parsing mode will be k.OBCIParsingNormal, and .streaming will be detected true or false depending on whether or not the board and dongle are already actually streaming prior to connection.

When a parameter of `true` is passed to disconnect(), streamStop() will not be called even if the board is streaming.

**Comment by aj-ptw** [comment]
I now understand the purpose of this PR.

Very interesting. I dig it. Good idea. Ok let's get these tests working!!! We will use spys! I will pull your pr and crank them out first thing tomorrow. Mine is AJ please to e-meet you! Thanks for your help. This is such a good idea.

File: openBCIBoard.js [hunk 1]
```diff
@@ -259,6 +260,22 @@ function OpenBCIFactory() {
...
                 var timeoutLength = this.options.simulate ? 50 : 300;
                 if(this.options.verbose) console.log('Serial port open');
+                if(dontReset) {
+                    this.streaming = false;
+                    this.curParsingMode = k.OBCIParsingNormal;
+                    var t, l;
+                    t = setTimeout(() => {
+                        this.removeListener('sample', l);
+                        resolve();
+                    },2 * 1000 / this.sampleRate());
```
Comment (aj-ptw) [inline openBCIBoard.js]: any reason this is two seconds? Could this be like 0.5 seconds or sample rate times two

**Comment by aj-ptw** [comment]
I've been working on it so just give me a little before making changes please!

**Comment by baffo32** [comment]
Hi my name is Karl.  Thank you for looking at this, if it's any trouble don't worry I'll get to it.  The timeout is actually sample rate converted from Hz to seconds, times two, so it is only some hundreds  of milliseconds.  - Sent from my phone

**Comment by aj-ptw** [comment]
gotcha! ok yup almost done writing the tests

**Comment by aj-ptw** [comment]
#85 might solve some issues with this pr

**Comment by aj-ptw** [comment]
The more I think about this PR, the more I think it should really be by default that we don't send things to the board.

**Comment by baffo32** [comment]
I agree, but that would break backwards compatibility.

**Comment by aj-ptw** [comment]
@baffo32 very excited to see if these tests pass!!!

**Comment by baffo32** [comment]
still a WIP but I think I made a lot of progress today.  I haven't enabled the `xit` tests yet.

**Comment by aj-ptw** [comment]
Ahh ok awesome

**Comment by aj-ptw** [comment]
@baffo32 ok i updated and merged #88 into a new branch called `1.4.0` as requested

File: openBCIBoard.js [hunk 2]
```diff
@@ -225,9 +225,9 @@ function OpenBCIFactory() {
      * @author AJ Keller (@pushtheworldllc)
      */
     OpenBCIBoard.prototype.connect = function(portName, dontReset) {
-        this.connected = false;
```
Comment (aj-ptw) [inline openBCIBoard.js]: agreed. this is a terrible idea to blindly set `connected` to `false`

File: openBCIBoard.js [hunk 3]
```diff
@@ -256,66 +261,110 @@ function OpenBCIFactory() {
...
+                        if(this.options.verbose) console.log("Waiting for '$$$'");
+                    },timeoutLength + 250);
+                }
             });
-            boardSerial.once('close',() => {
+
+            var closeListener = () => {
+                this.serial = null;
+
+                this.commandsToWrite = 0;
```
Comment (aj-ptw) [inline openBCIBoard.js]: I *love* the inclusion of the write function here!!!

File: openBCIBoard.js [hunk 4]
```diff
@@ -256,66 +261,110 @@ function OpenBCIFactory() {
...
+
+        // TODO: wait for commandsToWrite to drain before closing
+
         var timeout = 0;
-        if (this.streaming) {
-            this.streamStop();
-            if(this.options.verbose) console.log('stop streaming');
-            timeout = 15; // Avg time is takes for message to propagate
+        if(this.streaming) {
+            console.log(`dontStop is ${dontStop}`);
```
Comment (aj-ptw) [inline openBCIBoard.js]: Need to either remove this line or add a `if(this.options.verbose)`

**Comment by aj-ptw** [comment]
@baffo32 I am looking at your code and there still needs some work on `disconnect`

I'm going to rebase your most recent commit over `1.4.0` branch

**Comment by baffo32** [comment]
Yes, that was what I was going to do next.  I've been splitting the needed changes into issues so that they could be re-evaluated in light of any changes you made in 1.4.0, which I haven't reviewed yet.  (also the work here is split over a few different commits between you and me, not all of them labeled with regard to what they address).

I'm going to try to get `disconnect` going locally on the branch as-is, so that I understand any related issues, and then address whatever the state of `1.4.0` is.

**Comment by aj-ptw** [comment]
Yup my apologies, i will pr your don't reset branch with the rebased #82 over `1.4.0`

**Comment by aj-ptw** [comment]
@baffo32 please look at #94 and I plan to merge that with master to push repo to `1.3.3` then you can continue working and we can deal with semi-standard

**Comment by baffo32** [comment]
Mostly done.  Everything merges fine.

I'd be glad to redo this, now I've partitioned off all the parts needed, again on top of 1.4.0 .  I think standardizing the style is valuable.

**Comment by baffo32** [comment]
@aj-ptw Do you think rebasing off 1.3.3 (rather than using existing merge) is worth it?  If I were to go through and make a rebase work it seems it would be more valuable with 1.4.0

**Comment by aj-ptw** [comment]
You rebased off `1.4.0` already?

**Comment by baffo32** [comment]
No, this has been merged with 1.3.3 .  But I would like to rebase off 1.4.0 with your go-ahead.

I'm not very familiar with rebasing; I'm more a merger.  It would be simpler for me to just redo the work and give the commits proper labels.

**Comment by aj-ptw** [comment]
It's a disaster trying to rebase this off of `1.4.0`, i tried haha

**Comment by aj-ptw** [comment]
`1.4.1` could add #50 too

**Comment by baffo32** [comment]
I think it would make more sense to redo the work completely, which is not a problem now I've identified all the issues one runs into.  Alternatively `1.4.0` could be rebased off this.

#50 will probably need tests; it's not immediately ready.

The 1.4 series could also adjust the API to make this `dontReset` behavior the default, as you mentioned earlier.

**Comment by aj-ptw** [comment]
> I think it would make more sense to redo the work completely, which is not a problem now I've identified all the issues one runs into. Alternatively 1.4.0 could be rebased off this.

Ok then. Please let me rebase `1.4.0` off master and then you can _redo thew or completely_

**Comment by aj-ptw** [comment]
@baffo32 ok an hour later `1.4.0` branch is rebased and ready to be built on top of. Don't do work off `master` any more!

**Comment by baffo32** [comment]
ok

**Comment by aj-ptw** [comment]
This PR should be closed because of plan because #96 will contain the relevant code covered in this PR.

**Pull request closed without merging**